### ISSUE TITLE
Unexpected Results in PyTorch Tensor Operations with Python Scalars

### ISSUE THREAD

**Issue by tshadley** [pytorch/pytorch#114679]
### 🐛 Describe the bug

`torch.tensor(1,dtype=torch.int8) < 256` returns False instead of True
`torch.tensor(1, dtype=torch.int16)//32768)` returns -1 instead of 0

This looks like python scalars are converted to the left or right tensor operand type without promotion (causing overflow) in mixed operand expressions.  Is this expected behavior?  

Numpy balances by promoting operands to the largest operand type (following C/C++ balancing/promotion rules).

This is reproduced with the nightly build installed: `pip install --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cu120`

```python
import torch
import numpy

print(torch.tensor(1,dtype=torch.int8) < 256)     # False unexpected
print(256 < torch.tensor(1,dtype=torch.int8))     # True unexpected
print(torch.tensor(1,dtype=torch.int8) < torch.tensor(256,dtype=torch.int16))
print(torch.tensor(1,dtype=torch.int8).item() < 256)
print(numpy.array(1,dtype=numpy.int8) < 256)
print(256 < numpy.array(1,dtype=numpy.int8))

print(torch.tensor(1, dtype=torch.int16)//32768)  #-1 unexpected
print(torch.tensor(1, dtype=torch.int16)//torch.tensor(32768))
print(torch.tensor(1, dtype=torch.int16)//32767)  
print(numpy.array(1, dtype=numpy.int16)//32768)
print(numpy.array(1, dtype=numpy.int16)//numpy.array(32768))
```

Result:

```
tensor(False)
tensor(True)
tensor(True)
True
True
False
tensor(-1, dtype=torch.int16)
tensor(0)
tensor(0, dtype=torch.int16)
0
0
```

### Versions

Collecting environment information...
PyTorch version: 2.1.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.27.5
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-88-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA TITAN X (Pascal)
GPU 2: NVIDIA TITAN Xp

Nvidia driver version: 535.129.03
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             20
On-line CPU(s) list:                0-19
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Core(TM) i9-9820X CPU @ 3.30GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 10
Socket(s):                          1
Stepping:                           4
CPU max MHz:                        4200.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           6599.98
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          320 KiB (10 instances)
L1i cache:                          320 KiB (10 instances)
L2 cache:                           10 MiB (10 instances)
L3 cache:                           16.5 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-19
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT vulnerable

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] onnx==1.14.1
[pip3] onnxoptimizer==0.3.13
[pip3] onnxruntime==1.16.0
[pip3] open-clip-torch==2.20.0
[pip3] pytorch-lightning==1.9.4
[pip3] pytorch-metric-learning==2.3.0
[pip3] pytorch-ranger==0.1.1
[pip3] torch==2.1.1
[pip3] torch-audiomentations==0.11.0
[pip3] torch-optimizer==0.3.0
[pip3] torch-pitch-shift==1.2.4
[pip3] torch-stft==0.1.4
[pip3] torchaudio==2.1.1
[pip3] torchdiffeq==0.2.3
[pip3] torchinfo==1.8.0
[pip3] torchmetrics==1.2.0
[pip3] torchsde==0.2.5
[pip3] torchsummary==1.5.1
[pip3] torchview==0.2.6
[pip3] torchvision==0.16.1
[pip3] torchviz==0.0.2
[pip3] triton==2.1.0
[conda] Could not collect
